src/join.rs

```diff
@@ -114,6 +121,106 @@ impl<J: Join> std::iter::Iterator for JoinIter<J> {
     }
 }
 
+#[cfg(feature="parallel")]
+pub use self::par_join::*;
+/// `JoinParIter` is an `ParallelIterator` over a group of `Storages`.
+#[must_use]
+#[cfg(feature="parallel")]
+pub struct JoinParIter<J: Join>(J);
+#[cfg(feature="parallel")]
+mod par_join {
+    use std::marker::PhantomData;
+    use std::sync::Arc;
+    use std::cell::UnsafeCell;
+    use rayon::iter::ParallelIterator;
+    use rayon::iter::internal::{UnindexedProducer, UnindexedConsumer, Folder, bridge_unindexed};
+    use super::*;
+    use bitset::BitProducer;
+
+    impl<J> ParallelIterator for JoinParIter<J>
+      where
+        J: Join + Send,
+        J::Type: Send,
+        J::Value: Split + Send,
+        J::Mask: Send + Sync,
+    {
+        type Item = J::Type;
+
+        fn drive_unindexed<C>(self, consumer: C) -> C::Result
+            where C: UnindexedConsumer<Self::Item>
+        {
+            let (keys, values) = self.0.open();
+            let producer = BitProducer((&keys).iter());
+            let values = Arc::new(UnsafeCell::new(values));
+            bridge_unindexed(JoinProducer::<J>::new(producer, values), consumer)
+        }
+    }
+
+    struct JoinProducer<'a, 'b, J>
+      where
+        J: Join + Send,
+        J::Type: Send,
+        J::Value: 'a + Send,
+        J::Mask: 'b + Send + Sync,
+    {
+        keys: BitProducer<'b, J::Mask>,
+        values: Arc<UnsafeCell<J::Value>>,
```

Comment:
Yeah, this is unfortunate; we have to make the guarantee that every possible safe code, even if it's wrong (or unusual), can not cause undefined behavior. However, consider the following:

```rust
pub struct MyUnusualCollection<T> {
    access_counter: usize, // incremented on every access
    inner: Vec<T>,
}
```

What I wanted to show is that even though it is usually the case that you *could* access the storage with multiple mutable references given they don't access the same access, we are not allowed to make such an assumption.

I didn't give it a try, but if I were you I'd try to go with a separate trait, as already proposed earlier, and optionally allow a storage to provide a `split` method, which would be just the [`Vec::chunks`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.chunks) method for vectors, for example.